### DISCUSSION TITLE
Clarification on the scorer code.

### DISCUSSION THREAD

**Discussion by raotnameh** [speechbrain/speechbrain#2582] [Q&A]
Hi, 

Thank you for the awesome library. 

Is there any document or video I can go through to understand how to implement the scorer? If there is not any,  could the team please guide me for the same. I would be more than happy to write a document later. 

Thanks.

**Comment by Adel-Moumen** [comment]
Hello @raotnameh, 

Do you mind explaining a bit more what you'd like to do precisely ? 

Thanks. 
Adel

**Reply by raotnameh** [reply]
Hi Adel, 

I am trying to use speechbrain for an encoder and decoder model. 

As of now, I have successfully implemented the: 
1. Model
2. Forward and Backward pass on the training dataset. 


During the decoding, I am not able to understand the overall flow speechbrain uses. How can I modify my decoder to be used by the scorer. 
It would be very helpful, if you could walk me through the steps. 
How to use the files in "speechbrain.decoders.scorer" to implement the decoding part with greedy or beam search + n-gram or combined (ctc + n-gram) and a LM. I am not able to navigate the files and understand them. 

One example of a yaml defination is shown below.  It would be very helpful if you could help me understand. 

`

ctc_scorer: !new:speechbrain.decoders.scorer.CTCScorer
    eos_index: !ref <eos_index>
    blank_index: !ref <blank_index>
    ctc_fc: !ref <ctc_lin>


transformerlm_scorer: !new:speechbrain.decoders.scorer.TransformerLMScorer
    language_model: !ref <lm_model>
    temperature: 1.15

scorer_test_search: !new:speechbrain.decoders.scorer.ScorerBuilder
    full_scorers: [!ref <transformerlm_scorer>, !ref <ctc_scorer>]
    weights:
        ctc: !ref <ctc_weight_decode>
        transformerlm: !ref <lm_weight>

scorer_valid_search: !new:speechbrain.decoders.scorer.ScorerBuilder
    full_scorers: [!ref <ctc_scorer>]
    weights:
        ctc: !ref <ctc_weight_decode>

valid_search: !new:speechbrain.decoders.S2STransformerBeamSearcher
    modules: [!ref <Transformer>, !ref <seq_lin>]
    bos_index: !ref <bos_index>
    eos_index: !ref <eos_index>
    min_decode_ratio: !ref <min_decode_ratio>
    max_decode_ratio: !ref <max_decode_ratio>
    beam_size: !ref <valid_beam_size>
    using_eos_threshold: False
    length_normalization: True
    scorer: !ref <scorer_valid_search>

test_search: !new:speechbrain.decoders.S2STransformerBeamSearcher
    modules: [!ref <Transformer>, !ref <seq_lin>]
    bos_index: !ref <bos_index>
    eos_index: !ref <eos_index>
    min_decode_ratio: !ref <min_decode_ratio>
    max_decode_ratio: !ref <max_decode_ratio>
    beam_size: !ref <test_beam_size>
    temperature: 1.15
    using_eos_threshold: False
    length_normalization: True
    scorer: !ref <scorer_test_search>

`

Thanks
